graphql.go

```diff
@@ -146,16 +189,57 @@ func (c *Client) runWithJSON(ctx context.Context, req *Request, resp interface{}
 		return NewExecutionError(errors.Wrap(err, "reading body"))
 	}
 	c.logf("<< %s", buf.String())
-	if err := json.NewDecoder(&buf).Decode(&gr); err != nil {
-		return NewExecutionError(errors.Wrap(err, "decoding response"))
+
+	var gr *graphResponse
+	switch op.(type) {
+	case *Mutation:
+		var results struct {
+			Data map[string]graphMutationPayload
+		}
+
+		if err := json.NewDecoder(&buf).Decode(&results); err != nil {
+			return NewExecutionError(errors.Wrap(err, "decoding response"))
+		}
+		gr = &graphResponse{}
+
+		for _, result := range results.Data {
+			if !result.Successful {
+				messages := result.Messages
+				errors := make([]graphErr, len(messages))
+
+				for i, message := range messages {
+					errors[i] = graphErr{
+						Message: emptyOrString(message.Message),
+						Code:    message.Code,
+					}
+				}
+
+				gr.Errors = append(gr.Errors, errors...)
+			} else {
+				err := mapstructure.Decode(results.Data, &resp)
+				if err != nil {
+					return NewExecutionError(errors.Wrap(err, "decoding response"))
+				}
+			}
+			// The code above only supports payloads with a single mutation
+			break
+		}
+
+	default:
+		gr = &graphResponse{Data: resp}
+		if err := json.NewDecoder(&buf).Decode(&gr); err != nil {
+			return NewExecutionError(errors.Wrap(err, "decoding response"))
+		}
```

Comment:
I tried setting up this way but I was faced with the same issue, when decoding `resp` we need to pass the structure where the data must be decoded, which at compile time would be of the type `interface {}`, because of this, I couldn't access the attributes of a concrete struct for example `results.Data`. Similarly, in test we use `map[string]interface{}` but we could also use a struct, in one of the test examples, and afaik there is no way of doing a direct conversion  which means we needed to do some sort of reflection anyways.   I would have ended up implementing something very similar to what this lib delivers.
The thing that I'm not happy about is that we go through the json twice, first to decode it as a `mutationPayload` and a second time to populate the `resp` =\